Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/MessagesRepository.cs

```diff
@@ -37,45 +41,49 @@ public async Task<Message> FindPlain(MessageId id, CancellationToken cancellatio
         return await _readOnlyMessages.FirstWithId(id, cancellationToken);
     }
 
-    public async Task<DbPaginationResult<Message>> FindAll(PaginationFilter paginationFilter)
-    {
-        return await _readOnlyMessages.OrderAndPaginate(d => d.CreatedAt, paginationFilter);
-    }
-
     public async Task<MessageId> Add(Message message, CancellationToken cancellationToken)
     {
         var add = await _messages.AddAsync(message, cancellationToken);
         await _dbContext.SaveChangesAsync(cancellationToken);
+        _blobStorage.Add(_blobOptions.RootFolder, message.Id, message.Body);
+        await _blobStorage.SaveAsync();
         return add.Entity.Id;
     }
 
-    public Task<int> CountUnreceivedMessagesFromActiveIdentity(IdentityAddress sender, SendMessageCommandRecipientInformation recipientDto, CancellationToken cancellationToken)
+    public Task<int> CountUnreceivedMessagesFromSenderToReceiver(IdentityAddress sender, IdentityAddress recipient, CancellationToken cancellationToken)
     {
         return _readOnlyMessages
             .FromASpecificSender(sender)
-            .WithASpecificRecipientWhoDidNotReceiveTheMessage(recipientDto.Address)
+            .WithASpecificRecipientWhoDidNotReceiveTheMessage(recipient)
             .CountAsync(cancellationToken);
     }
 
-    public Task<DbPaginationResult<Message>> FindMessagesOfIdentity(IdentityAddress identityAddress, ListMessagesQuery request, IdentityAddress addressOfActiveIdentity, CancellationToken cancellationToken)
+    public Task<DbPaginationResult<Message>> FindMessagesWithIds(IEnumerable<MessageId> ids, IdentityAddress requiredParticipant, PaginationFilter paginationFilter, bool track = false)
     {
-        var query = _readOnlyMessages
+        var query = (track ? _messages : _readOnlyMessages)
             .AsQueryable()
             .IncludeAllReferences();
 
-        if (request.Ids.Any())
-            query = query.WithIdsIn(request.Ids);
+        if (ids.Any())
+            query = query.WithIdsIn(ids);
 
-        query = query.WithSenderOrRecipient(identityAddress);
-        query = query.DoNotSendBeforePropertyIsNotInTheFuture();
+        return query.WithSenderOrRecipient(requiredParticipant)
+            .DoNotSendBeforePropertyIsNotInTheFuture()
+            .OrderAndPaginate(d => d.CreatedAt, paginationFilter);
+    }
 
-        return query.OrderAndPaginate(d => d.CreatedAt, request.PaginationFilter);
+    public async Task Update(Message message)
+    {
+        _messages.Update(message);
+        await _dbContext.SaveChangesAsync();
     }
 
-    public EntityEntry<Message> Update(Message message)
+    public async Task FetchedMessage(Message message, IdentityAddress address, DeviceId deviceId)
     {
-        var m = _messages.Update(message);
-        _dbContext.SaveChanges();
-        return m;
+        var recipient = message.Recipients.FirstWithIdOrDefault(address);
+
+        recipient.ReceivedMessage(deviceId);
```

Comment:
You misunderstood me here. I meant that you should rename the `recipient.ReceivedMessage` method to `recipient.FetchedMessage` and then call it directly from the Handlers. After that, you just need to call Update for the Message(s).
And you also should add an `Update` overload that accepts multiple Messages. Because currently, when you mark 1,000 Messages as received, you are making 1,000 database requests, which performs poorly.

Reply:
Due to other requested changes, I've had to change the `Find*` methods return types as follows:
```C#
Task<ListMessagesResponse> FindMessagesWithIds(...);
Task<MessageDTO> Find(...);
```

with these, I cannot call the `recipient.FetchedMessage` considering I do not have access to the message(s).

Am I missing something here, or is this indeed a problem? How should we proceed? Thanks.

Reply:
Which requested changes do you mean? I definitely never intended your described change. The repositories should accept and return only aggregates - no DTOs, Commands, Queries or Responses.

Reply:
For instance, let's focus on the `Find` method. The repository should fill the body from the BlobStorage. However, the `Message.Body` attribute is read only (`{ get; }`).

Considering the controller returns a `MessageDTO`, which has a writable `Body`, I assumed it would be the best approach. My bad.

Should the `Message.Body` attribute be made writable instead?

Reply:
Oh, got it.

You should introduce I private setter for the `Body` property. And a method called `LoadBody`. This makes it explicit that it should only be used to load a body, and not to change it.

Inside of the method, you should additionally check if the Body is null. If it is not, throw an Exception.